### ISSUE TITLE
`overload`ed function doesn't accept a `Union` 

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#1119]
**Software I'm using**
Python 3.9, reproduced on both Pyright and Pylance extensions in VSCode.

**The bug**
OK:
```py
def make_dict(source: Iterable[Tuple[str, int]]):
    result = {}
    result.update(source)
    return result
```

OK:
```py
def make_dict(source: Mapping[str, int]):
    result = {}
    result.update(source)
    return result
```

NOT OK:
```py
def make_dict(source: Union[Mapping[str, int], Iterable[Tuple[str, int]]]):
    result = {}
    result.update(source)
    return result
```
Error is: 
```
No overloads for "result.update(source)" match parameters
  Argument types: (Mapping[str, int] | Iterable[Tuple[str, int]])
```
But it's actually correctly typed.
`dict.update` has two overloads: for `Mapping[str, int]` and for `Iterable[Tuple[str, int]]`, and both overloads return `None`. So really, `dict.update` is the same as a function that accepts `(Mapping[str, int] | Iterable[Tuple[str, int]])`. 

# Change in `typeshed` 
A comment for `typing.pyi` in `typeshed` says that `MutableMapping.update` was changed from a union to an overload, so that might have revealed the issue.

# Minimal reproducible example:
```py
@overload
def f(x: int) -> str:
    ...

@overload
def f(x: str) -> str:
    ...

def f(x):
    return str(x)


def g(x: Union[int, str]):
    return f(x)
```
Error:
```
No overloads for "f(x)" match parameters
  Argument types: (int | str)
```

**Comment by decorator-factory** [comment]
Possibly related to https://github.com/microsoft/pyright/issues/1088

Comment from `typeshed`:
```
    # 'update' used to take a Union, but using overloading is better.
    # The second overloaded type here is a bit too general, because
    # Mapping[Tuple[_KT, _VT], W] is a subclass of Iterable[Tuple[_KT, _VT]],
    # but will always have the behavior of the first overloaded type
    # at runtime, leading to keys of a mix of types _KT and Tuple[_KT, _VT].
    # We don't currently have any way of forcing all Mappings to use
    # the first overload, but by using overloading rather than a Union,
    # mypy will commit to using the first overload when the argument is
    # known to be a Mapping with unknown type parameters, which is closer
    # to the behavior we want. See mypy issue  #1430.
```

**Comment by erictraut** [comment]
This speaks to a minor difference in the way that mypy and pyright treat overloads with respect to unions. When mypy encounters an argument with a union type, it performs an overload match for each subtype in that union, perhaps choosing different overloads for each. Pyright performs only a single overload match. None of the PEPs specify this behavior, so it is not standardized. The reason I chose in pyright to do it this way is because 1) it rarely matters (as evidenced by the fact that this is the first time anyone has ever reported an issue related to it), 2) it's how overloads work in other languages like TypeScript, and 3) the other approach can lead to combinatoric behaviors that explode the analysis time.

I've thought about perhaps changing the behavior in pyright to match mypy, but it is a big structural change in the analyzer and could have some significant performance impacts.

**Comment by decorator-factory** [comment]
Thanks for the reply. In my case, I can just change the function interface to accept a mapping, and I might be overcomplicating things.

**Comment by erictraut** [comment]
Reopening this issue. We've received a sufficient number of bug reports about the overload matching behavior that I decided to change the behavior to more closely match mypy's. This eliminates the need for workarounds like the one I posted above. This change will be in the next version of pylance.

**Comment by erictraut** [comment]
This change is included in pyright 1.1.135, which I just published. It will also be included in the next release of pylance.

**Comment by decorator-factory** [comment]
Thanks :+1:
